b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,384 @@
+            self._mDNSRegistrationHandle = mdns.registerService(serviceInfo, null);
+          }
+
+          aResolve();
+          self._serverStatus = SERVER_STATUS.STARTED;
+        }
+      }
+    });
+  },
+
+  // Notifies this server that the given connection has been closed.
+  _connectionClosed: function(connection) {
+    DEBUG && debug("Close connection " + connection.connectionId);
+    this._connections.delete(connection.connectionId);
+  },
+};
+
+function streamClosed(e) {
+  return e === Cr.NS_BASE_STREAM_CLOSED ||
+         (typeof e === "object" && e.result === Cr.NS_BASE_STREAM_CLOSED);
+}
+
+// Represents a connection to the server
+function Connection(input, output, server, connectionId) {
+  DEBUG && debug("Open a new connection " + connectionId);
+
+  // Server associated with this connection
+  this.server = server;
+
+  // Id of this connection
+  this.connectionId = connectionId;
+
+  this.eventHandler = null;
+
+  // Input and output UTF-8 stream
+  this._input = Cc["@mozilla.org/intl/converter-input-stream;1"]
+                  .createInstance(Ci.nsIConverterInputStream);
+  this._input.init(input, "UTF-8", 0,
+                   Ci.nsIConverterInputStream.DEFAULT_REPLACEMENT_CHARACTER);
+
+  this._output = Cc["@mozilla.org/intl/converter-output-stream;1"]
+                   .createInstance(Ci.nsIConverterOutputStream);
+  this._output.init(output, "UTF-8", 0, 0x0000);
+
+  // This allows a connection to disambiguate between a peer initiating a
+  // close and the socket being forced closed on shutdown.
+  this._closed = false;
+}
+Connection.prototype = {
+  // Closes this connection's input/output streams
+  close: function() {
+    if (this._closed) {
+      return;
+    }
+
+    DEBUG && debug("Close connection " + this.connectionId);
+
+    this.input.close();
+    this.output.close();
+    this._closed = true;
+
+    this.server._connectionClosed(this);
+  },
+
+  // nsIInputStreamCallback
+  onInputStreamReady: function(input) {
+    DEBUG && debug("onInputStreamReady(input=" + input + ") on thread " +
+                   Services.tm.currentThread + " (main is " +
+                   Services.tm.mainThread + ")");
+
+    try {
+      let available = 0, numChars = 0, fullMessage = "";
+
+      // Read and concat messages from input stream buffer
+      do {
+        let partialMessage = {};
+
+        available = input.available();
+        numChars = this._input.readString(available, partialMessage);
+
+        fullMessage += partialMessage.value;
+      } while(numChars < available);
+
+      if (fullMessage.length > 0) { // While readString contains something
+        // Handle concatenated JSON string
+        fullMessage = this._sanitizeMessage(fullMessage);
+        debug(fullMessage);
+
+        try {
+          // Parse JSON string to event objects
+          let events = JSON.parse(fullMessage);
+
+          events.forEach((event) => {
+            this.eventHandler.handleEvent(event);
```

Comment:
check the existence of this.eventHandler, or you should pass the EventHandler in ctor.